File: javascript/tests/constraint.test.js
/**
 * Tests for constraint validation.
 */

import { test, describe } from 'node:test';
import assert from 'node:assert';
import { parseRateLimit, domainPatternMatches, constraintsSubsetOf } from '../src/constraint.js';

describe('parseRateLimit', () => {
    test('parses rates correctly', () => {
        assert.strictEqual(parseRateLimit('100/hour'), 100);
        assert.strictEqual(parseRateLimit('10/minute'), 600);
        assert.strictEqual(parseRateLimit('1/second'), 3600);
    });

    test('returns null for invalid input', () => {
        assert.strictEqual(parseRateLimit('invalid'), null);
        assert.strictEqual(parseRateLimit('100/day'), null);
    });
});

describe('domainPatternMatches', () => {
    test('exact match', () => {
        assert.ok(domainPatternMatches('example.com', 'example.com'));
    });

    test('wildcard match', () => {
        assert.ok(domainPatternMatches('*.example.com', 'sub.example.com'));
        assert.ok(!domainPatternMatches('*.example.com', 'example.com'));
    });

    test('no match', () => {
        assert.ok(!domainPatternMatches('other.com', 'example.com'));
    });
});

describe('constraintsSubsetOf', () => {
    test('valid subset', () => {
        const disc = {
            data_classification_max: 'confidential',
            rate_limit: '100/hour',
        };
        const cred = {
            data_classification_max: 'internal',
            rate_limit: '50/hour',
        };
        assert.ok(constraintsSubsetOf(disc, cred));
    });

    test('data classification exceeds', () => {
        const disc = { data_classification_max: 'confidential' };
        const cred = { data_classification_max: 'restricted' };
        assert.ok(!constraintsSubsetOf(disc, cred));
    });

    test('null discovery allows anything', () => {
        assert.ok(constraintsSubsetOf(null, { rate_limit: '1000/hour' }));
    });

    test('null credential is fine', () => {
        assert.ok(constraintsSubsetOf({ rate_limit: '100/hour' }, null));
    });
});

File: go/pkg/types/constraint_test.go
package types

import (
	"encoding/json"
	"testing"
)

func TestDataClassificationOrdering(t *testing.T) {
	if DataPublic.Order() >= DataInternal.Order() {
		t.Fatal("public should be < internal")
	}
	if DataInternal.Order() >= DataConfidential.Order() {
		t.Fatal("internal should be < confidential")
	}
	if DataConfidential.Order() >= DataRestricted.Order() {
		t.Fatal("confidential should be < restricted")
	}
}

func TestParseRateLimit(t *testing.T) {
	cases := []struct {
		in  string
		out uint64
		ok  bool
	}{
		{"100/hour", 100, true},
		{"10/minute", 600, true},
		{"1/second", 3600, true},
		{"bad", 0, false},
		{"100/year", 0, false},
	}
	for _, tc := range cases {
		got, ok := parseRateLimit(tc.in)
		if got != tc.out || ok != tc.ok {
			t.Errorf("parseRateLimit(%q) = (%d,%v), want (%d,%v)", tc.in, got, ok, tc.out, tc.ok)
		}
	}
}

func TestDomainPatternMatches(t *testing.T) {
	if !domainPatternMatches("example.com", "example.com") {
		t.Fatal("exact match")
	}
	if !domainPatternMatches("*.example.com", "sub.example.com") {
		t.Fatal("wildcard subdomain")
	}
	if domainPatternMatches("*.example.com", "example.com") {
		t.Fatal("wildcard does not match bare")
	}
	if domainPatternMatches("other.com", "example.com") {
		t.Fatal("different domains must not match")
	}
}

func TestConstraintsSubsetOf(t *testing.T) {
	conf := DataConfidential
	intl := DataInternal
	rest := DataRestricted

	disc := &Constraints{
		DataClassificationMax: &conf,
		RateLimit:             "100/hour",
	}
	credOK := &Constraints{
		DataClassificationMax: &intl,
		RateLimit:             "50/hour",
	}
	if !ConstraintsSubsetOf(disc, credOK) {
		t.Fatal("credOK should be a subset")
	}

	credBad := &Constraints{
		DataClassificationMax: &rest,
	}
	if ConstraintsSubsetOf(disc, credBad) {
		t.Fatal("restricted > confidential should fail")
	}
}

func TestConstraintsJSONRoundTrip(t *testing.T) {
	intl := DataInternal
	c := Constraints{
		AllowedDomains:        []string{"*.example.com"},
		RateLimit:             "50/hour",
		DataClassificationMax: &intl,
	}
	data, err := json.Marshal(c)
	if err != nil {
		t.Fatal(err)
	}
	var c2 Constraints
	if err := json.Unmarshal(data, &c2); err != nil {
		t.Fatal(err)
	}
	if c2.RateLimit != c.RateLimit || *c2.DataClassificationMax != *c.DataClassificationMax {
		t.Fatal("constraints roundtrip mismatch")
	}
}
